testament/testament.nim

```diff
@@ -1220,155 +1457,861 @@ proc parseArgs(execState: var Execution, p: var OptParser): ParseCliResult =
+  # actionId, along with a pair of `next` ones so we can serialize the compile
+  # and execute phase of compile program then run compiled program.
+  var
+    testCmds = newSeqOfCap[string](batchSize)
+    cmdIdToActId = newSeqOfCap[int](batchSize)
+    cmdIdToActKind = newSeqOfCap[TTestAction](batchSize)
+    cmdIdToInput = newSeqOfCap[Option[string]](batchSize)
+    nextTestCmds = newSeqOfCap[string](batchSize)
+    nextCmdIdToActId = newSeqOfCap[int](batchSize)
+    nextCmdIdToActKind = newSeqOfCap[TTestAction](batchSize)
+    nextCmdIdToInput = newSeqOfCap[Option[string]](batchSize)
+    exitCodes {.threadvar.}: seq[int]
+    outputs {.threadvar.}: seq[string]
+    startTimes {.threadvar.}: seq[float]
+    endTimes {.threadvar.}: seq[float]
+    batches = 0
+    startedCategoryIds: seq[CategoryId]
+
+  exitCodes = newSeq[int](batchSize)
+  outputs = newSeq[string](batchSize)
+  startTimes = newSeq[float](batchSize)
+  endTimes = newSeq[float](batchSize)
+
+  proc onTestRunStart(id: int) =
+    if verbose:
+      msg Undefined: "executing: " & testCmds[id]
+
+    startTimes[id] = epochTime()
+    # reset
+    exitCodes[id] = 0
+    outputs[id] = ""
+    endTimes[id] = 0.0
+
+  proc onTestProcess(id: int, p: Process) =
+    let testInput = cmdIdToInput[id]
+    if testInput.isSome:
+      let instream = inputStream(p)
+      instream.write(testInput.get())
+      close instream
     else:
-      let processOpts =
-        if optFailing and not optVerbose:
-          {poStdErrToStdOut, poUsePath, poParentStreams}
+      discard
+
+  proc onTestRunComplete(id: int, p: Process) =
+    if verbose:
+      msg Undefined: "finished execution of '$#' with code $#" % [
+        testCmds[id],
+        $p.peekExitCode()
+      ]
+
+    endTimes[id] = epochTime()
+    exitCodes[id] = p.peekExitCode()
+    let outp = p.outputStream
+    outputs[id] = outp.readAll
+    outp.close()
+
+  for actionId, action in testActions:
+    let
+      runId = action.runId
+      testRun = testRuns[runId]
+      testId = testRun.testId
+      spec = execState.testSpecs[testId]
+      testFile = execState.testFiles[testId]
+      matrixOptions =
+        if testRun.matrixEntry == noMatrixEntry:
+          ""
         else:
-          {poEchoCmd, poStdErrToStdOut, poUsePath, poParentStreams}
-      let qval = osproc.execProcesses(cmds, processOpts, beforeRunEvent = progressStatus)
-      if backendLogging:
-        backend.cacheResults()
-      quit qval
+          spec.matrix[testRun.matrixEntry]
+      target = testRun.target
+      nimcache = nimcacheDir(testFile.file, matrixOptions & testArgs, target)
+      compileCmd = prepareTestCompileCmd(
+        cmdTemplate = spec.getCmd,
+        filename = testFile.file,
+        options = testArgs,
+        nimcache = nimcache,
+        target = target,
+        extraOptions = matrixOptions,
+        outfile = testRun.exeFile(testFile.file, execState.rootDir))
+
```

Comment:
Continuous block or lets, the i visual break nterruptions had me checking to see if it's local assignment.

Reply:
I will break it into smaller `let` section

Reply:
### My Adivce

If you see a pattern that simplifies across the whole test loop then go for it, otherwise I'd recommend preserving the style:

1. establish facts/read queries
2. if necessary setup local mutables
3. do whatever logic and write out any updates
    - aim to mutate a 'row' of state only in one place (single writer)

### Philosphicaly

I recommend keeping it together, I know it's a smelly, but it's good to keep smelly code smelly. It's counter-intuitive, but the way we simplify the code is by thinking through these lets, what they're trying to do, if there is simplification or not. By keeping it smelly we keep the code honest, don't let it fake a clean look.

### My Intention at the Time

They're all acting as sort of facts/read queries of the state. I was aiming to figure out some Command Query Responsibility Segregation type pattern. Where reads (these queries) and the mutable bits (commands later one) are segregated.

With enough of these let blocks identified across various parts of the code, I was thinking it'd be possible to look across them for common factors. Then see if they're worth pulling together into some other abstraction, my hunch is was it might end up like this:

- have the exec state object similar to today
- have some query procs for common queries
- mutation for any given piece of state should only happen once per overall loop in one spot

But the precise decomposition I was waiting to see more before running ahead with it.